### ISSUE TITLE
exllama Mistral loras doesnt work.

### ISSUE THREAD

**Issue by noiraku** [oobabooga/text-generation-webui#4272]
### Describe the bug

Literally doesnt work i tried both exllama and exllamav2 with _HF ones as well.
However lora works with transformers but slow af... we really need exllama for this.

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Reproduction

-git pull latest version
-start_window.bat with nvidia choice
-add model TheBloke/Mistral-7B-Instruct-v0.1-GPTQ
-add lora that trained with (TheBloke/Mistral-7B-Instruct-v0.1-GPTQ) same model
-Apply lora.

### Logs

```shell
EXLLAMA:

2023-10-12 21:50:46 INFO:Loading TheBloke_Mistral-7B-Instruct-v0.1-GPTQ...
2023-10-12 21:50:48 INFO:Replaced attention with xformers_attention
2023-10-12 21:50:48 INFO:Loaded the model in 1.71 seconds.
2023-10-12 21:50:51 INFO:Applying the following LoRAs to TheBloke_Mistral-7B-Instruct-v0.1-GPTQ: noirakulora100


Common sense questions and answers

Question: test
Factual answer:
--------------------

Traceback (most recent call last):
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\text_generation.py", line 387, in generate_reply_custom
    for reply in shared.model.generate_with_streaming(question, state):
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\exllama.py", line 158, in generate_with_streaming
    token = self.generator.gen_single_token()
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\exllama\generator.py", line 362, in gen_single_token
    token, _ = self.batched_sample(logits,
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\exllama\generator.py", line 65, in batched_sample
    if logits.shape[0] == 1: return self.sample(logits, temperature, top_k, top_p, min_p, typical, num)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\exllama\generator.py", line 162, in sample
    sampled_ind = torch.multinomial(top_probs, top_probs.shape[-1] if num == -1 else min(num, top_probs.shape[-1]))
RuntimeError: probability tensor contains either `inf`, `nan` or element < 0
Output generated in 0.29 seconds (0.00 tokens/s, 0 tokens, context 16, seed 973931404)

--------------------------------------------------------------------------------------------------

EXLLAMAV2:
2023-10-12 21:52:07 INFO:Loading TheBloke_Mistral-7B-Instruct-v0.1-GPTQ...
2023-10-12 21:52:10 INFO:Replaced attention with xformers_attention
2023-10-12 21:52:10 INFO:Loaded the model in 2.82 seconds.
Traceback (most recent call last):
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\queueing.py", line 406, in call_prediction
    output = await route_utils.call_process_api(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\route_utils.py", line 226, in call_process_api
    output = await app.get_blocks().process_api(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\blocks.py", line 1554, in process_api
    result = await self.call_function(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\blocks.py", line 1206, in call_function
    prediction = await utils.async_iteration(iterator)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\utils.py", line 517, in async_iteration
    return await iterator.__anext__()
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\utils.py", line 510, in __anext__
    return await anyio.to_thread.run_sync(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\anyio\to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\anyio\_backends\_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\anyio\_backends\_asyncio.py", line 807, in run
    result = context.run(func, *args)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\utils.py", line 493, in run_sync_iterator_async
    return next(iterator)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\utils.py", line 647, in gen_wrapper
    yield from f(*args, **kwargs)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\ui_model_menu.py", line 222, in load_lora_wrapper
    add_lora_to_model(selected_loras)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\LoRA.py", line 17, in add_lora_to_model
    add_lora_transformers(lora_names)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\LoRA.py", line 127, in add_lora_transformers
    params['peft_type'] = shared.model.dtype
AttributeError: 'Exllamav2Model' object has no attribute 'dtype'

--------------------------------------------------------------------------------------------------

EXLLAMA_HF:
2023-10-12 21:52:55 INFO:Loading TheBloke_Mistral-7B-Instruct-v0.1-GPTQ...
2023-10-12 21:52:56 INFO:Replaced attention with xformers_attention
2023-10-12 21:52:56 INFO:Loaded the model in 1.63 seconds.
2023-10-12 21:53:00 INFO:Applying the following LoRAs to TheBloke_Mistral-7B-Instruct-v0.1-GPTQ: noirakulora100


Common sense questions and answers

Question: test
Factual answer:
--------------------

Traceback (most recent call last):
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\callbacks.py", line 56, in gentask
    ret = self.mfunc(callback=_callback, *args, **self.kwargs)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\text_generation.py", line 349, in generate_with_callback
    shared.model.generate(**kwargs)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\transformers\generation\utils.py", line 1652, in generate
    return self.sample(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\transformers\generation\utils.py", line 2770, in sample
    next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)
RuntimeError: probability tensor contains either `inf`, `nan` or element < 0
Output generated in 0.44 seconds (0.00 tokens/s, 0 tokens, context 16, seed 1693713814)

--------------------------------------------------------------------------------------------------

EXLLAMAv2_HF:

2023-10-12 21:53:42 INFO:Loading TheBloke_Mistral-7B-Instruct-v0.1-GPTQ...
2023-10-12 21:53:44 INFO:Replaced attention with xformers_attention
2023-10-12 21:53:44 INFO:Loaded the model in 2.62 seconds.
2023-10-12 21:53:47 INFO:Applying the following LoRAs to TheBloke_Mistral-7B-Instruct-v0.1-GPTQ: noirakulora100
Traceback (most recent call last):
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\queueing.py", line 406, in call_prediction
    output = await route_utils.call_process_api(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\route_utils.py", line 226, in call_process_api
    output = await app.get_blocks().process_api(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\blocks.py", line 1554, in process_api
    result = await self.call_function(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\blocks.py", line 1206, in call_function
    prediction = await utils.async_iteration(iterator)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\utils.py", line 517, in async_iteration
    return await iterator.__anext__()
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\utils.py", line 510, in __anext__
    return await anyio.to_thread.run_sync(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\anyio\to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\anyio\_backends\_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\anyio\_backends\_asyncio.py", line 807, in run
    result = context.run(func, *args)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\utils.py", line 493, in run_sync_iterator_async
    return next(iterator)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\gradio\utils.py", line 647, in gen_wrapper
    yield from f(*args, **kwargs)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\ui_model_menu.py", line 222, in load_lora_wrapper
    add_lora_to_model(selected_loras)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\LoRA.py", line 17, in add_lora_to_model
    add_lora_transformers(lora_names)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\modules\LoRA.py", line 134, in add_lora_transformers
    shared.model = PeftModel.from_pretrained(shared.model, get_lora_path(lora_names[0]), adapter_name=lora_names[0], **params)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\peft\peft_model.py", line 277, in from_pretrained
    model = MODEL_TYPE_TO_PEFT_MODEL_MAPPING[config.task_type](model, config, adapter_name)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\peft\peft_model.py", line 889, in __init__
    super().__init__(model, peft_config, adapter_name)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\peft\peft_model.py", line 111, in __init__
    self.base_model = PEFT_TYPE_TO_MODEL_MAPPING[peft_config.peft_type](
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\peft\tuners\lora.py", line 274, in __init__
    super().__init__(model, config, adapter_name)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\peft\tuners\tuners_utils.py", line 88, in __init__
    self.inject_adapter(self.model, adapter_name)
  File "J:\NAIVOIC\OOBABOOGALLM\text-generation-webui00666fff\text-generation-webui\installer_files\env\lib\site-packages\peft\tuners\tuners_utils.py", line 222, in inject_adapter
    raise ValueError(
ValueError: Target modules ['q_proj', 'v_proj'] not found in the base model. Please check the target modules and try again.
```

### System Info

```shell
Win10
Nvidia 3080 10gb
```

**Comment by nbollman** [comment]
I don't think you are having problems loading the Mistral model, I think you are getting the errors trying to load the LORA on the model?
Me: Id like to summarize the root cause of these errors, do they occur when the Mistral-7b model is loaded or when the LORA is loaded?

ClaudeAI:
Based on the tracebacks provided, it seems the errors occur when trying to load and apply the LORA (LoRAAdapter) to the Mistral-7B model.

Specifically, the errors appear to occur in these main steps:

Loading the Mistral-7B model - This seems to succeed without errors.
Applying the LORA - This is where the errors start occurring:
For the EXLLAMA and EXLLAMA_HF versions, the error occurs when trying to get the 'dtype' attribute of the model, which doesn't seem to exist.
For the EXLLAMAv2_HF version, the error occurs when trying to inject the LORA adapter into the model. Specifically, it cannot find the target modules 'q_proj' and 'v_proj' in the base Mistral-7B model.
So in summary, loading the Mistral-7B model itself succeeds, but applying the LORA adapter to the model fails in different ways depending on the Python library being used (EXLLAMA vs EXLLAMAv2_HF). The root cause seems to be incompatibility between the LORA adapter code and the base Mistral-7B model structure/interfaces.

**Comment by nbollman** [comment]
Sorry, I must have misread the title of your post, I thought it read 'exllama Mistral doesnt work' my tired eyes missed the lora, apologies for the stupid answer.
